src/header/dib_header.rs

```diff
@@ -0,0 +1,138 @@
+        // bytes to get the right final offset to the end of the header.
+        let (input, dib_header_data) =
+            length_data(map_opt(le_u32, |len| len.checked_sub(4)))(input)?;
+
+        // Add 4 back on so the constants remain the correct size relative to the BMP
+        // documentation/specs.
+        let header_type = match dib_header_data.len() + 4 {
+            DIB_V3_HEADER_SIZE => HeaderType::V3,
+            DIB_V4_HEADER_SIZE => HeaderType::V4,
+            DIB_V5_HEADER_SIZE => HeaderType::V5,
+            DIB_INFO_HEADER_SIZE => HeaderType::Info,
+            _ => {
+                return Err(nom::Err::Failure(nom::error::Error::from_error_kind(
+                    dib_header_data,
+                    ErrorKind::LengthValue,
+                )))
+            }
+        };
+
+        // Fields common to all DIB variants
+        let (dib_header_data, image_width) = le_u32(dib_header_data)?;
+        let (dib_header_data, image_height) = le_i32(dib_header_data)?;
+        let (dib_header_data, _color_planes) = le_u16(dib_header_data)?;
+        let (dib_header_data, bpp) = Bpp::parse(dib_header_data)?;
+
+        // Extra fields defined by DIB variants
+        // Variants are described in
+        // <https://www.liquisearch.com/bmp_file_format/file_structure/dib_header_bitmap_information_header>
+        // and <https://docs.microsoft.com/en-us/windows/win32/gdi/bitmap-header-types>
+        let (dib_header_data, compression_method) = CompressionMethod::parse(dib_header_data)?;
+        let (dib_header_data, image_data_len) = le_u32(dib_header_data)?;
+        let (dib_header_data, _pels_per_meter_x) = le_u32(dib_header_data)?;
+        let (dib_header_data, _pels_per_meter_y) = le_u32(dib_header_data)?;
+        let (dib_header_data, colors_used) = le_u32(dib_header_data)?;
+        let (dib_header_data, colors_important) = le_u32(dib_header_data)?;
+
+        let (_dib_header_data, channel_masks) = if header_type.is_at_least(HeaderType::V3)
+            && compression_method == CompressionMethod::Bitfields
+        {
+            let (dib_header_data, mask_red) = le_u32(dib_header_data)?;
+            let (dib_header_data, mask_green) = le_u32(dib_header_data)?;
+            let (dib_header_data, mask_blue) = le_u32(dib_header_data)?;
+            let (dib_header_data, mask_alpha) = le_u32(dib_header_data)?;
+
+            (
+                dib_header_data,
+                Some(ChannelMasks {
+                    red: mask_red,
+                    green: mask_green,
+                    blue: mask_blue,
+                    alpha: mask_alpha,
+                }),
+            )
+        } else {
+            (dib_header_data, None)
+        };
+
+        // Number of colors in the color table. If the specific count is zero, the entire color
+        // space should be used.
+        let color_table_num_entries: u32 = if colors_important == 0 {
```

Comment:
I've added this check as part of the recent review changes. Both [here](https://docs.microsoft.com/en-us/windows/win32/api/wingdi/ns-wingdi-bitmapv4header) and [here](https://www.liquisearch.com/bmp_file_format/file_structure/dib_header_bitmap_information_header) are quite vague about this field - the latter even go as far as saying it's "generally ignored" - however if I don't do this check, various bits of our test suite fail. 

This check feels like a kludge because I don't have any firm evidence of it being the correct logic, but our tests pass with it like this so...

Reply:
If I understand the purpose of the `colors_important` field correctly it is used to define the minimum number of different colors that are necessary to display a reasonable representation of the image. But this is probably an ancient feature that was used to distinguish between images that could be displayed on an EGA display or that need one of those fancy VGA ones. We should ignore this field.

I assume the reasons the tests fail is the value of `color_table_num_entries`. The docs are very vague, but I believe that a `0` in `colors_used` translates to `color_table_num_entries == 0` for >= 16 bpp.